### PR TITLE
log-viewer-webui: Add server implementation for submitting IR extraction jobs and serving IR files; Don't copy unnecessary files when building component.

### PR DESCRIPTION
**Pull request by wraymo**
# Description

This PR adds a basic server implementation for log-viewer-webui.
+ It includes a DbManager to handle connections for both MySQL and MongoDB.
+ It adds routes for querying and inserting data.

# Validation performed

+ Built and started the clp package, added `CLP_DB_PASS` and `CLP_DB_USER` to `.env` file.
+ Ran `npm test`.

### CODE REVIEW

File: components/log-viewer-webui/server/src/app.test.js
Comment (junhaoliao) [inline components/log-viewer-webui/server/src/app.test.js:9]: We might want to update the [docs](https://docs.yscope.com/clp/main/dev-guide/components-log-viewer-webui.html#testing) to state the CLP package / testing DB servers should be up before these tests can be run.

**Review by junhaoliao — COMMENTED** [review]
Thanks a lot for picking up my backlog. Let's update some of the implementations after #460 is merged.

File: components/log-viewer-webui/server/src/DbManager.js
Comment (haiqi96) [inline components/log-viewer-webui/server/src/DbManager.js:80]: why `id`? shouldn't it be type?

File: components/log-viewer-webui/server/src/app.js
Comment (junhaoliao) [inline components/log-viewer-webui/server/src/app.js:17]:
```suggestion
```

File: components/log-viewer-webui/server/.env
Comment (junhaoliao) [inline components/log-viewer-webui/server/.env:8]:
```suggestion
CLP_DB_PASS=

```

File: components/log-viewer-webui/server/src/app.js
Comment (junhaoliao) [inline components/log-viewer-webui/server/src/app.js:13]:
```suggestion
 * Creates the Fastify app with the given options.
```

File: components/log-viewer-webui/server/.env
Comment (junhaoliao) [inline components/log-viewer-webui/server/.env:3]: @Henry8192 Kindly update here once your `yscope-log-viewer` integration PR is ready.

File: components/clp-package-utils/clp_package_utils/scripts/start_clp.py
Comment (junhaoliao) [inline components/clp-package-utils/clp_package_utils/scripts/start_clp.py:801]: We should also add `LOG_VIEWER_DIR` once that is ready.

File: components/log-viewer-webui/server/src/app.js
Comment (junhaoliao) [inline components/log-viewer-webui/server/src/app.js:55]: Needed for any second and after registration of fastifyStatic: https://github.com/fastify/fastify-static#disabling-reply-decorator

File: components/log-viewer-webui/server/src/app.js
Comment (junhaoliao) [inline components/log-viewer-webui/server/src/app.js:71]: This is needed to avoid decorating reply twice with fastifyStatic, which yields an error: https://github.com/fastify/fastify-static#disabling-reply-decorator

File: components/log-viewer-webui/server/src/app.test.js
Comment (junhaoliao) [inline components/log-viewer-webui/server/src/app.test.js:7]: @kirkrodrigues 
For routes that have dependencies on external data source (e.g., databases), how do we want to test those?
Shall we implement a stub class of DbManager?

File: Taskfile.yml
Comment (junhaoliao) [inline Taskfile.yml:203]: The plural form here is to prepare for yscope-log-viewer UI integration, which will be in another PR by @Henry8192 .

File: Taskfile.yml
Comment (junhaoliao) [inline Taskfile.yml:97]: hey @haiqi96 - when you're available, could you help reviewing the Taskfile and CLP packaging script changes in this PR? There are some changes that might be against what we previously discussed (e.g., we wanted to copy the source files into `<clp-root>/build` before building the client assets); however, with some new measures added (e.g., an additional argument passed to `npm run build`), some previous conclusions might no longer apply. Feel free to object.

Comment (junhaoliao) [reply Taskfile.yml:97]: @kirkrodrigues / @haiqi96 : Let me know if I should split the submission into two PRs (one for log-viewer-webui and one for packaging) now that I realize this one is kinda long.

File: Taskfile.yml
Comment (haiqi96) [inline Taskfile.yml:150]: Just wonder, will the code still work if we first run npm clean-install in another directory and copy what's built into var/www/log_viewer_webui?

File: components/clp-package-utils/clp_package_utils/scripts/start_clp.py
Comment (haiqi96) [inline components/clp-package-utils/clp_package_utils/scripts/start_clp.py:796]: I think you are overwriting every fields in the settings.json. in that case, do we still need to require the file to exist?

File: Taskfile.yml
Comment (junhaoliao) [reply Taskfile.yml:150]: I think so. However sometimes within node_modules there're symbolic links created during the node dependencies installation, so it's less ideal to copy an initialized node_modules.

File: components/clp-package-utils/clp_package_utils/scripts/start_clp.py
Comment (junhaoliao) [reply components/clp-package-utils/clp_package_utils/scripts/start_clp.py:796]: I remember discussing a similar situation with @kirkrodrigues when we added the meteor settings.json override but I forgot the conclusion. Do you remember?

File: components/clp-package-utils/clp_package_utils/scripts/start_clp.py
Comment (junhaoliao) [inline components/clp-package-utils/clp_package_utils/scripts/start_clp.py:780]: We can add `log_viewer_webui_logs_dir` back once logging to file is enabled in log-viewer-webui-server.

File: components/log-viewer-webui/server/src/routes/query.js
Comment (kirkrodrigues) [inline components/log-viewer-webui/server/src/routes/query.js:32]:
```suggestion
        return irMetadata;
```

File: components/log-viewer-webui/server/src/app.test.js
Comment (kirkrodrigues) [inline components/log-viewer-webui/server/src/app.test.js:8]: Instead of forcibly setting the environment, maybe we should assert that it's "test"?

File: components/log-viewer-webui/server/src/main.js
Comment (kirkrodrigues) [inline components/log-viewer-webui/server/src/main.js:77]: Do we still need this export?

File: components/log-viewer-webui/server/settings.json
Comment (kirkrodrigues) [inline components/log-viewer-webui/server/settings.json:12]: How about `IrFilesDir`?

File: components/log-viewer-webui/server/src/DbManager.js
Comment (kirkrodrigues) [inline components/log-viewer-webui/server/src/DbManager.js:68]:
```suggestion
```

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]
Review for the webserver part.

File: components/log-viewer-webui/server/src/app.test.js
Comment (kirkrodrigues) [reply components/log-viewer-webui/server/src/app.test.js:7]: I know a lot of people will use mocked databases, etc. but I feel like those introduce unnecessary complexity and ultimately aren't a very good substitute for actually talking to a database directly. So my preference would be to separate the tests which rely on a database from those that don't, and for the former, we can start up a database before running those tests. We could add some scripts to the repo to start-up a database container specifically for testing.

File: components/clp-package-utils/clp_package_utils/scripts/start_clp.py
Comment (kirkrodrigues) [reply components/clp-package-utils/clp_package_utils/scripts/start_clp.py:796]: The idea is that `settings.json` is an authoritative list of the settings that can exist. If we try to set a setting that doesn't exist, `update_settings_json` will complain.

Comment (kirkrodrigues) [reply components/clp-package-utils/clp_package_utils/scripts/start_clp.py:796]: To make that clearer, perhaps we should load `settings.json` inside `update_settings_json`.

File: Taskfile.yml
Comment (kirkrodrigues) [inline Taskfile.yml:97]: This task should also depend on the log-viewer-webui-clients checksum as well as the server source files, right?

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]
Ready for revisions.

File: Taskfile.yml
Comment (kirkrodrigues) [reply Taskfile.yml:203]: I'd prefer to leave it to Henry's PR.

File: components/clp-package-utils/clp_package_utils/scripts/start_clp.py
Comment (junhaoliao) [reply components/clp-package-utils/clp_package_utils/scripts/start_clp.py:801]: We do. That was accidentally removed. Sorry

File: components/log-viewer-webui/server/package.json
Comment (junhaoliao) [reply components/log-viewer-webui/server/package.json:21]: Although the version name has "beta" in it, the release is marked as an official release on npmjs registry and has been released for a year. 

https://www.npmjs.com/package/@msgpack/msgpack?activeTab=versions

File: components/log-viewer-webui/server/package.json
Comment (junhaoliao) [reply components/log-viewer-webui/server/package.json:29]: I tried replacing it with `@babel/plugin-syntax-import-attributes` and found the replacement was harmless (ESLint was still able to parse the import JSON line in app.js). Let's keep it as `@babel/plugin-syntax-import-attributes` then.

File: components/log-viewer-webui/server/src/app.test.js
Comment (junhaoliao) [reply components/log-viewer-webui/server/src/app.test.js:8]: In fact this NODE_ENV is unset if we don't have the line. However, I do agree the setting of it should be done in the package.json (we should specify it when running `tap`).

File: components/log-viewer-webui/server/src/routes/example.js
Comment (junhaoliao) [reply components/log-viewer-webui/server/src/routes/example.js:9]: The `await`s seem unnecessary as I found the route registrations are synchronous already. Will remove those all together.

File: components/log-viewer-webui/server/src/app.test.js
Comment (junhaoliao) [reply components/log-viewer-webui/server/src/app.test.js:7]: Sure. Can we do that in another PR? I can leave a TODO in this test file.

File: components/log-viewer-webui/server/package.json
Comment (kirkrodrigues) [reply components/log-viewer-webui/server/package.json:21]: Software. 🤦‍♂️

File: Taskfile.yml
Comment (kirkrodrigues) [reply Taskfile.yml]: We also want to add the server source files as dependencies, right?

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

**Review by kirkrodrigues — APPROVED** [review]
For the PR title, how about:

log-viewer-webui: Add server implementation for submitting IR extraction jobs and serving IR files; Don't copy unnecessary files when building component.

**Pull request merged**
